File: java/src/hotel/Reservation.java
package hotel;

import java.time.LocalDateTime;
import java.time.Duration;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Scanner;

/**
 * ╔═══════════════════════════════════════════════════════════════════════════╗
 * ║                    CLASSE RESERVATION (COEUR DU PROJET)                    ║
 * ╠═══════════════════════════════════════════════════════════════════════════╣
 * ║  Gestion d'une réservation avec ArrayList<Chambre>                        ║
 * ║  Maximum 80 chambres par hôtel                                            ║
 * ║  États: en_attente, validee, annulee                                      ║
 * ║  Projet EMSI - Préparé par Oussama SAJJI                                  ║
 * ╚═══════════════════════════════════════════════════════════════════════════╝
 */
public class Reservation {
    
    // ═══════════════════════════════════════════════════════════════════════════
    // CONSTANTES
    // ═══════════════════════════════════════════════════════════════════════════
    
    public static final int MAX_CHAMBRES = 80;
    public static final String ETAT_EN_ATTENTE = "en_attente";
    public static final String ETAT_VALIDEE = "validee";
    public static final String ETAT_ANNULEE = "annulee";
    
    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm");
    
    // ═══════════════════════════════════════════════════════════════════════════
    // ATTRIBUTS
    // ═══════════════════════════════════════════════════════════════════════════
    
    private int id;
    private String code;
    private int nbChambres;
    private ArrayList<Chambre> chambres;
    private Client client;
    private String etat;
    private LocalDateTime dateDebut;
    private LocalDateTime dateFin;
    
    // ═══════════════════════════════════════════════════════════════════════════
    // CONSTRUCTEURS
    // ═══════════════════════════════════════════════════════════════════════════
    
    /**
     * Constructeur par défaut
     */
    public Reservation() {
        this.id = 0;
        this.code = "";
        this.nbChambres = 0;
        this.chambres = new ArrayList<>();
        this.client = null;
        this.etat = ETAT_EN_ATTENTE;
        this.dateDebut = LocalDateTime.now();
        this.dateFin = LocalDateTime.now().plusDays(1);
    }
    
    /**
     * Constructeur avec code et client
     * @param code Code unique de la réservation
     * @param client Client effectuant la réservation
     */
    public Reservation(String code, Client client) {
        this();
        this.code = code;
        this.client = client;
    }
    
    /**
     * Constructeur complet
     * @param id Identifiant unique
     * @param code Code de réservation
     * @param client Client
     * @param dateDebut Date de début
     * @param dateFin Date de fin
     */
    public Reservation(int id, String code, Client client, LocalDateTime dateDebut, LocalDateTime dateFin) {
        this.id = id;
        this.code = code;
        this.nbChambres = 0;
        this.chambres = new ArrayList<>();
        this.client = client;
        this.etat = ETAT_EN_ATTENTE;
        this.dateDebut = dateDebut;
        this.dateFin = dateFin;
    }
    
    // ═══════════════════════════════════════════════════════════════════════════
    // GETTERS
    // ═══════════════════════════════════════════════════════════════════════════
    
    public int getId() { return id; }
    public String getCode() { return code; }
    public int getNbChambres() { return nbChambres; }
    public ArrayList<Chambre> getChambres() { return chambres; }
    public Client getClient() { return client; }
    public String getEtat() { return etat; }
    public LocalDateTime getDateDebut() { return dateDebut; }
    public LocalDateTime getDateFin() { return dateFin; }
    
    // ═══════════════════════════════════════════════════════════════════════════
    // SETTERS
    // ═══════════════════════════════════════════════════════════════════════════
    
    public void setId(int id) { this.id = id; }
    public void setCode(String code) { this.code = code; }
    public void setClient(Client client) { this.client = client; }
    public void setEtat(String etat) { this.etat = etat; }
    public void setDateDebut(LocalDateTime dateDebut) { this.dateDebut = dateDebut; }
    public void setDateFin(LocalDateTime dateFin) { this.dateFin = dateFin; }
    
    // ═══════════════════════════════════════════════════════════════════════════
    // METHODES PRINCIPALES (selon le PDF)
    // ═══════════════════════════════════════════════════════════════════════════
    
    /**
     * Vérifie si une chambre est déjà réservée dans cette réservation
     * @param chambre Chambre à vérifier
     * @return true si la chambre est déjà dans la réservation
     */
    public boolean verifierChambreReservee(Chambre chambre) {
        for (Chambre c : chambres) {
            if (c.equals(chambre)) {
                return true;
            }
        }
        return false;
    }
    
    /**
     * Ajoute une chambre à la réservation
     * @param chambre Chambre à ajouter
     * @return true si ajout réussi, false si max atteint ou doublon
     */
    public boolean ajouterChambre(Chambre chambre) {
        // Vérifier la limite de 80 chambres
        if (chambres.size() >= MAX_CHAMBRES) {
            System.out.println("⚠️ Limite de " + MAX_CHAMBRES + " chambres atteinte!");
            return false;
        }
        
        // Vérifier si la chambre est déjà réservée
        if (verifierChambreReservee(chambre)) {
            System.out.println("⚠️ La chambre " + chambre.getNumero() + " est déjà dans cette réservation!");
            return false;
        }
        
        // Ajouter la chambre
        chambres.add(chambre);
        nbChambres++;
        System.out.println("✅ Chambre " + chambre.getNumero() + " ajoutée à la réservation.");
        return true;
    }
    
    /**
     * Supprime une chambre de la réservation
     * @param chambre Chambre à supprimer
     */
    public void supprimerChambre(Chambre chambre) {
        boolean supprimee = false;
        for (int i = 0; i < chambres.size(); i++) {
            if (chambres.get(i).equals(chambre)) {
                chambres.remove(i);
                nbChambres--;
                supprimee = true;
                System.out.println("✅ Chambre " + chambre.getNumero() + " supprimée de la réservation.");
                break;
            }
        }
        
        if (!supprimee) {
            System.out.println("⚠️ Chambre non trouvée dans cette réservation.");
        }
    }
    
    /**
     * Calcule la durée de la réservation en heures
     * @return Durée en heures
     */
    public long calculerDuree() {
        if (dateDebut == null || dateFin == null) {
            return 0;
        }
        return Duration.between(dateDebut, dateFin).toHours();
    }
    
    /**
     * Calcule la durée de la réservation en jours
     * @return Durée en jours
     */
    public long calculerDureeJours() {
        long heures = calculerDuree();
        return (heures + 23) / 24; // Arrondi supérieur
    }
    
    /**
     * Valide la réservation
     */
    public void valider() {
        if (etat.equals(ETAT_ANNULEE)) {
            System.out.println("⚠️ Impossible de valider une réservation annulée.");
            return;
        }
        this.etat = ETAT_VALIDEE;
        System.out.println("✅ Réservation " + code + " validée.");
    }
    
    /**
     * Annule la réservation
     */
    public void annuler() {
        this.etat = ETAT_ANNULEE;
        System.out.println("❌ Réservation " + code + " annulée.");
    }
    
    /**
     * Vérifie si la réservation est expirée (date de fin dépassée)
     * @return true si expirée
     */
    public boolean estExpiree() {
        return dateFin != null && dateFin.isBefore(LocalDateTime.now());
    }
    
    /**
     * Vérifie si la réservation commence aujourd'hui
     * @return true si commence aujourd'hui
     */
    public boolean commenceAujourdhui() {
        if (dateDebut == null) return false;
        LocalDateTime maintenant = LocalDateTime.now();
        return dateDebut.toLocalDate().equals(maintenant.toLocalDate());
    }
    
    /**
     * Vérifie si la réservation se termine aujourd'hui
     * @return true si se termine aujourd'hui
     */
    public boolean seTermineAujourdhui() {
        if (dateFin == null) return false;
        LocalDateTime maintenant = LocalDateTime.now();
        return dateFin.toLocalDate().equals(maintenant.toLocalDate());
    }
    
    // ═══════════════════════════════════════════════════════════════════════════
    // SAISIE ET AFFICHAGE
    // ═══════════════════════════════════════════════════════════════════════════
    
    /**
     * Saisie interactive d'une réservation
     * @param scanner Scanner pour la lecture
     */
    public void saisir(Scanner scanner) {
        System.out.println("\n--- Saisie Réservation ---");
        
        System.out.print("Code de réservation: ");
        this.code = scanner.nextLine();
        
        System.out.print("Date début (jj/mm/aaaa HH:mm): ");
        String dateDebutStr = scanner.nextLine();
        try {
            this.dateDebut = LocalDateTime.parse(dateDebutStr, FORMATTER);
        } catch (Exception e) {
            System.out.println("Format invalide, utilisation de la date actuelle.");
            this.dateDebut = LocalDateTime.now();
        }
        
        System.out.print("Date fin (jj/mm/aaaa HH:mm): ");
        String dateFinStr = scanner.nextLine();
        try {
            this.dateFin = LocalDateTime.parse(dateFinStr, FORMATTER);
        } catch (Exception e) {
            System.out.println("Format invalide, utilisation de date début + 1 jour.");
            this.dateFin = this.dateDebut.plusDays(1);
        }
    }
    
    /**
     * Affichage des informations de la réservation
     */
    public void afficher() {
        System.out.println("\n╔══════════════════════════════════════════════════════════╗");
        System.out.println("║               RESERVATION " + String.format("%-30s", code) + " ║");
        System.out.println("╠══════════════════════════════════════════════════════════╣");
        System.out.println("║  ID: " + String.format("%-50d", id) + " ║");
        System.out.println("║  Client: " + String.format("%-46s", (client != null ? client.toString() : "N/A")) + " ║");
        System.out.println("║  État: " + String.format("%-48s", getEtatAffichage()) + " ║");
        System.out.println("║  Début: " + String.format("%-47s", dateDebut.format(FORMATTER)) + " ║");
        System.out.println("║  Fin: " + String.format("%-49s", dateFin.format(FORMATTER)) + " ║");
        System.out.println("║  Durée: " + String.format("%-47s", calculerDureeJours() + " jour(s)") + " ║");
        System.out.println("║  Nb Chambres: " + String.format("%-41d", nbChambres) + " ║");
        System.out.println("╠══════════════════════════════════════════════════════════╣");
        System.out.println("║  Chambres réservées:                                     ║");
        
        if (chambres.isEmpty()) {
            System.out.println("║    (aucune chambre)                                      ║");
        } else {
            for (Chambre c : chambres) {
                System.out.println("║    - " + String.format("%-50s", c.toString()) + " ║");
            }
        }
        
        System.out.println("╚══════════════════════════════════════════════════════════╝");
    }
    
    /**
     * Retourne l'état avec emoji
     * @return État formaté
     */
    private String getEtatAffichage() {
        switch (etat) {
            case ETAT_EN_ATTENTE: return "⏳ En attente";
            case ETAT_VALIDEE: return "✅ Validée";
            case ETAT_ANNULEE: return "❌ Annulée";
            default: return etat;
        }
    }
    
    @Override
    public String toString() {
        return "Réservation " + code + " - " + (client != null ? client.toString() : "N/A") + 
               " - " + nbChambres + " chambre(s) - " + etat;
    }
}

File: src/lib/classes/Reservation.js
/* eslint-disable @typescript-eslint/no-require-imports */
/**
 * Classe Reservation - Projet EMSI par Oussama SAJJI
 * 
 * Une reservation est identifiee par :
 * - code unique
 * - nombre de chambres reservees
 * - donnees des chambres reservees
 * - client qui a effectue la reservation
 * - etat de la reservation (EnCours, Validee, Annulee, Expiree)
 * - date de debut et fin (format: jour, mois, annee, heure)
 */

const Client = require("./Client");
const Chambre = require("./Chambre");

// Nombre maximum de chambres autorisees par reservation
const MAX_CHAMBRES = 5;

class Reservation {
  // Constructeur par defaut
  constructor(
    id = null,
    code = "",
    client = null,
    etat = "EnCours",
    dateDebut = null,
    dateFin = null,
    chambres = []
  ) {
    this._id = id;
    this._code = code;
    this._client = client; // Instance de Client
    this._etat = etat; // EnCours, Validee, Annulee, Expiree
    this._dateDebut = dateDebut ? new Date(dateDebut) : null;
    this._dateFin = dateFin ? new Date(dateFin) : null;
    this._chambres = chambres; // Array d'instances Chambre
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // GETTERS
  // ═══════════════════════════════════════════════════════════════════════════

  get id() {
    return this._id;
  }

  get code() {
    return this._code;
  }

  get client() {
    return this._client;
  }

  get etat() {
    return this._etat;
  }

  get dateDebut() {
    return this._dateDebut;
  }

  get dateFin() {
    return this._dateFin;
  }

  get chambres() {
    return this._chambres;
  }

  get nombreChambres() {
    return this._chambres.length;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // SETTERS
  // ═══════════════════════════════════════════════════════════════════════════

  set id(value) {
    this._id = value;
  }

  set code(value) {
    this._code = value;
  }

  set client(value) {
    this._client = value;
  }

  set etat(value) {
    this._etat = value;
  }

  set dateDebut(value) {
    this._dateDebut = value ? new Date(value) : null;
  }

  set dateFin(value) {
    this._dateFin = value ? new Date(value) : null;
  }

  set chambres(value) {
    this._chambres = value;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // METHODES DE LA CLASSE RESERVATION (selon cahier des charges)
  // ═══════════════════════════════════════════════════════════════════════════

  /**
   * a) Verifie si une chambre est deja reservee dans cette reservation
   * @param {Chambre} chambre - La chambre a verifier
   * @returns {boolean} True si la chambre est deja reservee, False sinon
   */
  verifierChambreReservee(chambre) {
    if (!(chambre instanceof Chambre)) {
      throw new Error("Le parametre doit etre une instance de Chambre");
    }
    return this._chambres.some((c) => c.comparer(chambre));
  }

  /**
   * b) Ajoute une chambre a la reservation
   * - Ne pas depasser le nombre de chambres autorisees (MAX_CHAMBRES)
   * - Rejeter l'ajout d'une chambre existante
   * @param {Chambre} chambre - La chambre a ajouter
   * @returns {Object} {success: boolean, message: string}
   */
  ajouterChambre(chambre) {
    if (!(chambre instanceof Chambre)) {
      return { success: false, message: "Le parametre doit etre une instance de Chambre" };
    }

    // Verifier la limite de chambres
    if (this._chambres.length >= MAX_CHAMBRES) {
      return {
        success: false,
        message: `Limite atteinte: impossible d'ajouter plus de ${MAX_CHAMBRES} chambres`,
      };
    }

    // Verifier si la chambre existe deja
    if (this.verifierChambreReservee(chambre)) {
      return {
        success: false,
        message: `La chambre ${chambre.numero} est deja dans cette reservation`,
      };
    }

    // Ajouter la chambre
    this._chambres.push(chambre);
    return {
      success: true,
      message: `Chambre ${chambre.numero} ajoutee avec succes`,
    };
  }

  /**
   * c) Affiche une reservation (informations des chambres et du client)
   * @returns {string}
   */
  afficher() {
    const formatDate = (d) => {
      if (!d) return "N/A";
      const jour = d.getDate().toString().padStart(2, "0");
      const mois = (d.getMonth() + 1).toString().padStart(2, "0");
      const annee = d.getFullYear();
      const heure = d.getHours().toString().padStart(2, "0");
      const minutes = d.getMinutes().toString().padStart(2, "0");
      return `${jour}/${mois}/${annee} ${heure}:${minutes}`;
    };

    const etatEmoji = {
      EnCours: "⏳",
      Validee: "✅",
      Annulee: "❌",
      Expiree: "⌛",
    };

    let result = `
╔═══════════════════════════════════════════════════════════════════════════╗
║                        RESERVATION: ${this._code.padEnd(20)}              ║
╠═══════════════════════════════════════════════════════════════════════════╣
║ Etat        : ${etatEmoji[this._etat] || ""} ${this._etat.padEnd(20)}                            ║
║ Date debut  : ${formatDate(this._dateDebut).padEnd(30)}                   ║
║ Date fin    : ${formatDate(this._dateFin).padEnd(30)}                   ║
║ Duree       : ${this.calculerDuree()} jour(s)                                           ║
╠═══════════════════════════════════════════════════════════════════════════╣
║ CLIENT                                                                    ║
╠───────────────────────────────────────────────────────────────────────────╣`;

    if (this._client) {
      result += `
║ ID       : ${String(this._client.id).padEnd(20)}                                  ║
║ Nom      : ${this._client.nom.padEnd(20)}                                  ║
║ Prenom   : ${this._client.prenom.padEnd(20)}                                  ║
║ Adresse  : ${this._client.adresse.padEnd(20)}                                  ║`;
    } else {
      result += `
║ Aucun client associe                                                       ║`;
    }

    result += `
╠═══════════════════════════════════════════════════════════════════════════╣
║ CHAMBRES RESERVEES (${this._chambres.length}/${MAX_CHAMBRES})                                           ║
╠───────────────────────────────────────────────────────────────────────────╣`;

    if (this._chambres.length === 0) {
      result += `
║ Aucune chambre reservee                                                   ║`;
    } else {
      this._chambres.forEach((ch, i) => {
        result += `
║ ${(i + 1)}. Chambre ${ch.numero.padEnd(10)} - Tel: ${ch.telephone.padEnd(15)}                 ║`;
      });
    }

    result += `
╚═══════════════════════════════════════════════════════════════════════════╝`;

    return result;
  }

  /**
   * d) Calcule la duree d'une reservation (en jours)
   * @returns {number} Nombre de jours
   */
  calculerDuree() {
    if (!this._dateDebut || !this._dateFin) return 0;
    const diffTime = Math.abs(this._dateFin.getTime() - this._dateDebut.getTime());
    const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
    return Math.max(1, diffDays); // Minimum 1 jour
  }

  /**
   * e) Supprime une chambre de la reservation
   * Affiche un message d'erreur si la chambre n'est pas reservee
   * @param {Chambre} chambre - La chambre a supprimer
   * @returns {Object} {success: boolean, message: string}
   */
  supprimerChambre(chambre) {
    if (!(chambre instanceof Chambre)) {
      return { success: false, message: "Le parametre doit etre une instance de Chambre" };
    }

    // Verifier si la chambre est dans la reservation
    if (!this.verifierChambreReservee(chambre)) {
      return {
        success: false,
        message: `ERREUR: La chambre ${chambre.numero} n'est pas dans cette reservation`,
      };
    }

    // Supprimer la chambre
    this._chambres = this._chambres.filter((c) => !c.comparer(chambre));
    return {
      success: true,
      message: `Chambre ${chambre.numero} supprimee avec succes`,
    };
  }

  /**
   * f) Valide la reservation (modifier l'etat)
   * @returns {Object} {success: boolean, message: string}
   */
  valider() {
    if (this._etat === "Annulee") {
      return { success: false, message: "Impossible de valider une reservation annulee" };
    }
    if (this._etat === "Validee") {
      return { success: false, message: "La reservation est deja validee" };
    }
    this._etat = "Validee";
    return { success: true, message: `Reservation ${this._code} validee avec succes` };
  }

  /**
   * f) Annule la reservation (modifier l'etat)
   * @returns {Object} {success: boolean, message: string}
   */
  annuler() {
    if (this._etat === "Annulee") {
      return { success: false, message: "La reservation est deja annulee" };
    }
    this._etat = "Annulee";
    return { success: true, message: `Reservation ${this._code} annulee avec succes` };
  }

  /**
   * Methode de saisie - remplit les attributs de la reservation
   * @param {Object} donnees
   */
  saisir(donnees) {
    if (donnees.id !== undefined) this._id = donnees.id;
    if (donnees.code !== undefined) this._code = donnees.code;
    if (donnees.client !== undefined) this._client = donnees.client;
    if (donnees.etat !== undefined) this._etat = donnees.etat;
    if (donnees.dateDebut !== undefined) this._dateDebut = new Date(donnees.dateDebut);
    if (donnees.dateFin !== undefined) this._dateFin = new Date(donnees.dateFin);
    if (donnees.chambres !== undefined) this._chambres = donnees.chambres;
  }

  /**
   * Compare deux reservations par leur code
   * @param {Reservation} autreReservation
   * @returns {boolean}
   */
  comparer(autreReservation) {
    if (!(autreReservation instanceof Reservation)) return false;
    return this._code === autreReservation.code;
  }

  /**
   * Convertit en objet simple
   * @returns {Object}
   */
  toObject() {
    return {
      id: this._id,
      code: this._code,
      clientId: this._client ? this._client.id : null,
      etat: this._etat,
      dateDebut: this._dateDebut,
      dateFin: this._dateFin,
      chambres: this._chambres.map((c) => c.toObject()),
    };
  }

  /**
   * Cree une instance depuis un objet Prisma
   * @param {Object} obj
   * @returns {Reservation}
   */
  static fromObject(obj) {
    const client = obj.client ? Client.fromObject(obj.client) : null;
    const chambres = obj.chambres
      ? obj.chambres.map((rc) => Chambre.fromObject(rc.chambre || rc))
      : [];

    return new Reservation(
      obj.id,
      obj.code,
      client,
      obj.etat,
      obj.dateDebut,
      obj.dateFin,
      chambres
    );
  }

  /**
   * Genere un code unique pour une nouvelle reservation
   * @returns {string}
   */
  static genererCode() {
    return `RES-${Date.now().toString(36).toUpperCase()}`;
  }
}

// Exporter la constante aussi
Reservation.MAX_CHAMBRES = MAX_CHAMBRES;

module.exports = Reservation;
